### PR TITLE
feat: apply c gwc api to rust bindings

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR involves a series of follow-up tasks after #296. 

- This fixes bugs w.r.t GWC prover.
- This adds rust bindings for GWC prover.

### CODE REVIEW

File: tachyon/c/zk/plonk/halo2/kzg_family_prover_impl.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+ private:
+  static CJacobianPoint* DoMSM(const std::vector<AffinePoint>& bases,
+                               const std::vector<ScalarField>& scalars) {
+    using MSM = tachyon::math::VariableBaseMSM<AffinePoint>;
+    MSM msm;
+    typename MSM::Bucket bucket;
+    absl::Span<const AffinePoint> bases_span = absl::Span<const AffinePoint>(
+        bases.data(), std::min(bases.size(), scalars.size()));
```
Comment (TomTaehoonKim) [inline tachyon/c/zk/plonk/halo2/kzg_family_prover_impl.h:50]:
```suggestion
    absl::Span<const AffinePoint> bases_span(
        bases.data(), std::min(bases.size(), scalars.size()));
```

File: vendors/halo2/src/prover.rs [hunk 2]
```diff
@@ -67,17 +74,36 @@ pub fn create_proof<'params, W: Write, ConcreteCircuit: Circuit<Fr>>(
...
-                        // NOTE(chokobole): P::QUERY_INSTANCE is removed since this isn't compiled well with it.
-                        // See https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/plonk/prover.rs#L91.
-                        transcript.common_scalar(values[i])?;
-                        poly.set_value(i, &values[i]);
+                        if !P::QUERY_INSTANCE {
+                            transcript.common_scalar(values[i])?;
+                        }
+                        poly.set_value(i, unsafe {
+                            std::mem::transmute::<_, &halo2curves::bn256::Fr>(&values[i])
+                        });
```
Comment (TomTaehoonKim) [inline vendors/halo2/src/prover.rs:82]: Is this change necessary?

File: vendors/halo2/src/prover.rs [hunk 3]
```diff
@@ -365,13 +391,18 @@ pub fn create_proof<'params, W: Write, ConcreteCircuit: Circuit<Fr>>(
...
+                    &mut advice_commitments,
+                );
                 let advice_commitments = advice_commitments;
                 drop(advice_commitments_projective);
 
                 for commitment in &advice_commitments {
-                    transcript.write_point(*commitment)?;
+                    transcript.write_point(unsafe {
+                        std::mem::transmute::<_, Scheme::Curve>(*commitment)
+                    })?;
```
Comment (TomTaehoonKim) [inline vendors/halo2/src/prover.rs:405]: Is this change necessary?

File: tachyon/zk/lookup/halo2/prover_impl.h [hunk 4]
```diff
@@ -193,6 +193,9 @@ void Prover<Poly, Evals>::Evaluate(ProverBase<PCS>* prover,
 #define EVALUATE(polynomial, point) \
   prover->EvaluateAndWriteToProof(polynomial.poly(), point_set.point)
 
+  // NOTE(chokobole): The ORDER IS IMPORTANT!!!
```
Comment (TomTaehoonKim) [inline tachyon/zk/lookup/halo2/prover_impl.h:196]:
```suggestion
  // The ORDER IS IMPORTANT!!!
```
I thought we were removing NOTE(author) comments. https://github.com/kroma-network/tachyon/commit/98db4aa004b6202d9ca0660b74456f30f1dc2102

File: tachyon/zk/lookup/halo2/prover_impl.h [hunk 5]
```diff
@@ -223,17 +226,20 @@ void Prover<Poly, Evals>::Open(
   base::Ref<const Poly>(&polynomial.poly()), point##_ref, \
       polynomial.poly().Evaluate(point_set.point)
 
+  // NOTE(chokobole): The ORDER IS IMPORTANT!!!
```
Comment (TomTaehoonKim) [inline tachyon/zk/lookup/halo2/prover_impl.h:229]:
```suggestion
  // The ORDER IS IMPORTANT!!!
```
I thought we were removing NOTE(author) comments. https://github.com/kroma-network/tachyon/commit/98db4aa004b6202d9ca0660b74456f30f1dc2102

File: tachyon/zk/plonk/permutation/permutation_prover_impl.h [hunk 6]
```diff
@@ -119,6 +119,9 @@ template <typename PCS>
 void PermutationProver<Poly, Evals>::Evaluate(
     ProverBase<PCS>* prover,
     const PermutationOpeningPointSet<F>& point_set) const {
+  // NOTE(chokobole): The ORDER IS IMPORTANT!!!
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/permutation/permutation_prover_impl.h:122]:
```suggestion
  // The ORDER IS IMPORTANT!!!
```
I thought we were removing NOTE(author) comments. https://github.com/kroma-network/tachyon/commit/98db4aa004b6202d9ca0660b74456f30f1dc2102

File: tachyon/zk/plonk/permutation/permutation_prover_impl.h [hunk 7]
```diff
@@ -158,6 +161,9 @@ void PermutationProver<Poly, Evals>::Open(
 #define OPENING(poly, point) \
   base::Ref<const Poly>(&poly), point##_ref, poly.Evaluate(point_set.point)
 
+  // NOTE(chokobole): The ORDER IS IMPORTANT!!!
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/permutation/permutation_prover_impl.h:164]:
```suggestion
  // The ORDER IS IMPORTANT!!!
```
I thought we were removing NOTE(author) comments. https://github.com/kroma-network/tachyon/commit/98db4aa004b6202d9ca0660b74456f30f1dc2102

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: vendors/halo2/src/prover.rs (hunk 3 above)
Comment (chokobole) [reply vendors/halo2/src/prover.rs:405]: Nice catch!

File: vendors/halo2/src/prover.rs (hunk 2 above)
Comment (chokobole) [reply vendors/halo2/src/prover.rs:82]: This is the error message without this.

```
error[E0308]: mismatched types
   --> vendors/halo2/src/prover.rs:80:43
    |
80  |                         poly.set_value(i, &values[i]);
    |                              ---------    ^^^^^^^^^^ expected struct `halo2curves::bn256::Fr`, found associated type
    |                              |
    |                              arguments to this function are incorrect
    |
    = note: expected reference `&halo2curves::bn256::Fr`
               found reference `&<Scheme as CommitmentScheme>::Scalar`
    = help: consider constraining the associated type `<Scheme as CommitmentScheme>::Scalar` to `halo2curves::bn256::Fr`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
note: associated function defined here
   --> vendors/halo2/src/bn254.rs:403:12
    |
403 |     pub fn set_value(&mut self, idx: usize, fr: &halo2curves::bn256::Fr) {
    |            ^^^^^^^^^                        ---------------------------

error: aborting due to previous error
```

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: vendors/halo2/src/bn254_gwc_prover.cc [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+  return std::make_unique<RationalEvals>(
+      tachyon_bn254_univariate_evaluation_domain_empty_rational_evals(
+          tachyon_halo2_bn254_gwc_prover_get_domain(prover_)));
+}
+
+std::unique_ptr<Poly> GWCProver::ifft(const Evals& evals) const {
+  // NOTE(chokobole): The zero degrees might be removed. This might cause an
+  // unexpected error if you use this carelessly. Since this is only used to
+  // compute instance polynomial and this is used only in Tachyon side, so it's
+  // fine.
```
Comment (ashjeong) [inline vendors/halo2/src/bn254_gwc_prover.cc:54]:
```suggestion
  // NOTE(chokobole): Leading zero values may be removed at this point, so use
  // this function cautiously. Tachyon currently uses it safely.

**Review by ashjeong — APPROVED** [review]
LGTM :)

**Pull request merged**